### ISSUE TITLE
Visualizing and Interpreting RNN Models in URL-based Phishing Detection

### ISSUE THREAD

**Issue by AndrewBreyen** [team3-cs373/resources#21]
### Title
Visualizing and Interpreting RNN Models in URL-based Phishing Detection

### URL
https://doi.org/10.1145/3381991.3395602

### Summary 
URLs are one of the main entry points to phishing attacks. Phishing detection such as Google Safe Browsing or Microsoft Smart Screen uses these URLs to protect users from these attacks. Serving up false positives may prohibit a user from visiting a website, but a false negative may expose the user to a potentially harmful attack. 

In order to make these systems more accurate, machine learning and heuristics can be used to increase the effectiveness of the filters. Heuristics of the URL itself is extremely effective without the need to analyze the website contents, up to 99% accuracy. Things like lexical features -- textual features and the number of n-grams, as well as host features such as the domain, registration, and location characteristics of the hosting server. 

A traditional ML approach to this issue using lexical features achieved a 97.3% accuracy percentage, showing that while effective, it could be better. 

Ma et all. used a Naive Bayes classification, and by using both lexical and host features, this resulted in an overall detection frequency of 95-99%. 

There are several other studies that found detection up to 99% accuracy, but the issue with the standard ML approach is that this relies on extraction of relevant features by researchers which requires manual effort and needs to be done repeatedly. 

When moving to the deep learning approach, no manual extraction is needed. There are many less studies on the deep learning approach, but it was still effective, about 98.7%. The RNN model in this study was iterated upon by these researchers and achieved a 99.2% accuracy rate. 

These researchers built four different URL based phishing detectors, LSTM, GRU, BiLSTM (bi-directional LSTM), and BiGRU (bi-directional GRU). 

Into these RNNs, a classifier was given (Ei= {e i0, · · · , e ij, · · · , e in−1}) and the output was the probability score of the URL being a phishing URL. If the output was >0.5, it was a phishing URL, else, it was not. 

When inputing the dataset of 1.5 million URLs, divided among training and validation, the RNNs were able to be trained within 25 epochs to an accuracy of over 99% and loss of less than 0.01. 

When analyzing false positive/negative rates, there was 1053 false positives and 392 false negatives over all four RNNs. These false positives were likely attributed to the URLs being short (gradeup.co) and false negatives likely being caused due to not containing enough characters to analyze. (iphaa.com)

The researchers also analyzed if having http:// vs http***S***:// matters to lexicographical phishing detection. They changed 30000 samples from http -> https, and from https -> http. For the four RNNs, the recall of all four decreased when http is changed to https: (LSTM: -10.4%, GRU: -10.3%, BiLSTM:-10.3%, BiGRU: -5.8%), indicating the corresponding obvious increase of false negative decisions) and when https is changed to http. " the Precision of all the four RNN models dropped significantly (LSTM:-48.1%, GRU: -34.6%, BiLSTM: -54.3%, BiGRU: -49.4%), indicating the corresponding significant increase of false positive decisions."

These results indicate that HTTPS vs HTTP is an important factor. However, as more and more phishing attacks switch to HTTPS, this may be less of an important factor. 

The RNNs also analyzed if a URL is an IP address or not, if there are special characters, different parts such as the protocol, host, path, and query string. 

The researchers also asked if RNNs are able to identify useful new features beyond what is used in a traditional phishing detection neural network. 

Conclusion: 
FOur RNNs were built using only lexical features of URLs to detect phishing. Unique visualizations were also developed for IP address detection and overall impact. The four RNN models do have a similar accuracy, there are differences in their accuracy. By combining all four RNNs, most false positives and false negatives could be eliminated. The RNNs are effective at extracting data that would normally be used in a "normal" neural network. 
### Citation
@inproceedings{10.1145/3381991.3395602,
author = {Feng, Tao and Yue, Chuan},
title = {Visualizing and Interpreting RNN Models in URL-Based Phishing Detection},
year = {2020},
isbn = {9781450375689},
publisher = {Association for Computing Machinery},
address = {New York, NY, USA},
url = {https://doi.org/10.1145/3381991.3395602},
doi = {10.1145/3381991.3395602},
abstract = {Existing studies have demonstrated that using traditional machine learning techniques, phishing detection simply based on the features of URLs can be very effective. In this paper, we explore the deep learning approach and build four RNN (Recurrent Neural Network) models that only use lexical features of URLs for detecting phishing attacks. We collect 1.5 million URLs as the dataset and show that our RNN models can achieve a higher than 99% detection accuracy without the need of any expert knowledge to manually identify the features. However, it is well known that RNNs and other deep learning techniques are still largely in black boxes. Understanding the internals of deep learning models is important and highly desirable to the improvement and proper application of the models. Therefore, in this work, we further develop several unique visualization techniques to intensively interpret how RNN models work internally in achieving the outstanding phishing detection performance. Especially, we identify and answer six important research questions, showing that our four RNN models (1) are complementary to each other and can be combined into an ensemble model with even better accuracy, (2) can well capture the relevant features that were manually extracted and used in the traditional machine learning approach for phishing detection, and (3) can help identify useful new features to enhance the accuracy of the traditional machine learning approach. Our techniques and experience in this work could be helpful for researchers to effectively apply deep learning techniques in addressing other real-world security or privacy problems.},
booktitle = {Proceedings of the 25th ACM Symposium on Access Control Models and Technologies},
pages = {13–24},
numpages = {12},
keywords = {phishing detection, visualization, deep learning, interpretation},
location = {Barcelona, Spain},
series = {SACMAT '20}
}
